apps/bfd-data-npi/src/main/java/gov/cms/bfd/data/npi/utility/DataUtilityCommons.java

```diff
@@ -0,0 +1,311 @@
+      if (!entry.isDirectory()) {
+        // if the entry is a file, extracts it
+        extractFile(zipIn, filePath);
+      } else {
+        // if the entry is a directory, make the directory
+        File dir = new File(filePath.toFile().getAbsolutePath());
+        dir.mkdir();
+      }
+      zipIn.closeEntry();
+      entry = zipIn.getNextEntry();
+    }
+    zipIn.close();
+  }
+
+  /**
+   * Extracts a zip entry (file entry).
+   *
+   * @param zipIn the zip file coming in
+   * @param filePath the file path for the file
+   * @throws IOException (any errors encountered will be bubbled up)
+   */
+  public static void extractFile(ZipInputStream zipIn, Path filePath) throws IOException {
+    BufferedOutputStream bos =
+        new BufferedOutputStream(new FileOutputStream(filePath.toFile().getAbsolutePath()));
+    byte[] bytesIn = new byte[BUFFER_SIZE];
+    int read = 0;
+    while ((read = zipIn.read(bytesIn)) != -1) {
+      bos.write(bytesIn, 0, read);
+    }
+    bos.close();
+  }
+
+  /**
+   * Creates the file in the specified location.
+   *
+   * @param workingDir a directory that temporary/working files can be written to
+   * @param fileName the output file/resource to produce
+   * @throws IOException (any errors encountered will be bubbled up)
+   * @return path to file
+   */
+  public static Path getOriginalNpiDataFile(Path workingDir, String fileName)
+      throws IOException, IllegalStateException {
+    // download NPI file
+    Path downloadedNpiZipFile =
+        Paths.get(workingDir.resolve("npidata.zip").toFile().getAbsolutePath());
+    URL ndctextZipUrl = new URL(fileName);
+    if (!Files.isReadable(downloadedNpiZipFile)) {
+      // connectionTimeout, readTimeout = 10 seconds
+      FileUtils.copyURLToFile(
+          ndctextZipUrl, new File(downloadedNpiZipFile.toFile().getAbsolutePath()), 100000, 100000);
+    }
+
+    // unzip NPI file.  Zip file contains these files
+    // pl_pfile_20050523-20220410.csv
+    // pl_pfile_20050523-20220410_FileHeader.csv
+    // othername_pfile_20050523-20220410.csv
+    // othername_pfile_20050523-20220410_FileHeader.csv
+    // NPPES_Data_Dissemination_Readme.pdf
+    // NPPES_Data_Dissemination_CodeValues.pdf
+    // npidata_pfile_20050523-20220410.csv
+    // npidata_pfile_20050523-20220410_FileHeader.csv
+    // endpoint_pfile_20050523-20220410.csv
+    // endpoint_pfile_20050523-20220410_FileHeader.csv
+    unzip(downloadedNpiZipFile, workingDir);
+    File f = new File(workingDir.toString());
+    File[] matchingFiles =
+        f.listFiles(
+            new FilenameFilter() {
+              public boolean accept(File dir, String name) {
+                return name.startsWith("npidata_pfile_") && !name.endsWith("_FileHeader.csv");
+              }
+            });
+
+    if (matchingFiles.length > 1) {
+      throw new IllegalStateException("More than one NPI file found");
+    }
```

Comment:
Should we also error if 0 files are found (say, if the filename prefix changes?)

Reply:
done